### PR TITLE
Dev fope

### PR DESCRIPTION
**Pull request by jayhenry**

### CODE REVIEW

File: tests/model/test_qwen3_moe.py [hunk 1]
```diff
@@ -339,7 +341,148 @@ def test_save_hf(self, device, dispatcher, ep_size):
...
+
+    fsdp_config = FSDPConfig(
+        ep_size=ep_size,
+        cpu_offload=False,
+    )
+    qwen_model.fully_shard(fsdp_config=fsdp_config)
+    qwen_model.from_hf(load_from)
+    return qwen_model
+
+def check_dict_equal(dict1: dict, dict2: dict) -> bool:
```
Comment (HAOCHENYE) [inline tests/model/test_qwen3_moe.py:468]: Why not use `self.assertDictEqual` in `TestCase`

File: tests/model/test_qwen3_moe.py [hunk 2]
```diff
@@ -339,7 +341,148 @@ def test_save_hf(self, device, dispatcher, ep_size):
...
+    
+    @parametrize.parametrize(
+        "device,dispatcher,ep_size",
+        [
+            ("cuda", None, 1),
+            ("cuda", "all2all", 4),
+            ("cuda", "all2all", 8),
+        ],
+    )
+    def test_save_hf_fope(self, device, dispatcher, ep_size):
```
Comment (HAOCHENYE) [inline tests/model/test_qwen3_moe.py:362]: Should we add accuracy test for FOPE?

File: tests/module/test_rope.py
Comment (HAOCHENYE) [inline tests/module/test_rope.py:1]: test_fope.py?

File: xtuner/v1/engine/train_engine.py [hunk 3]
```diff
@@ -207,7 +207,7 @@ def train_step(self, data_batches: list[ModelItem]):
         """
         if self.float8_handler is not None and self.float8_handler.enabled:
             self.float8_handler.precompute_float8_dynamic_scale_for_fsdp(self.model)
-
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/engine/train_engine.py]: clean the code

File: xtuner/v1/engine/train_engine.py [hunk 4]
```diff
@@ -262,6 +262,7 @@ def train_step(self, data_batches: list[ModelItem]):
                 total_forward_tokens += (num_tokens.sum()) ** 2
 
             if self.intra_layer_micro_batch == 1:
+                # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/engine/train_engine.py]: clean the code

File: xtuner/v1/model/base.py [hunk 5]
```diff
@@ -196,7 +201,9 @@ def from_hf(self, hf_path: str | Path, strict: bool = True) -> tuple:
             hf_path = str(hf_path)
 
         hf_loader = HFCheckpointLoader(hf_path)
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/base.py]: clean the code

File: xtuner/v1/model/base.py [hunk 6]
```diff
@@ -196,7 +201,9 @@ def from_hf(self, hf_path: str | Path, strict: bool = True) -> tuple:
...
         loaded_keys, unloaded_keys, missing_keys = self._load_params(hf_loader, strict=strict)
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/base.py]: clean the code

File: xtuner/v1/model/base.py [hunk 7]
```diff
@@ -933,7 +952,9 @@ def _load_same_hf_param(
     ) -> list[str]:  # return missing key
         local_tensor = param._local_tensor if isinstance(param, DTensor) else param
         hf_key = load_spec.hf_keys[0]
-
+        # TODO: 加载 sin_coef
+        # if 'sin_coef' in hf_key:
+        #     torch.distributed.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/base.py]: clean the code

File: xtuner/v1/model/base.py [hunk 8]
```diff
@@ -942,6 +963,8 @@ def _load_same_hf_param(
             loaded_tensor = self._load_fp8(hf_key, checkpoint_loader)
         else:
             loaded_tensor = checkpoint_loader.load(hf_key)
+        # if 'sin_coef' in hf_key:
+        #     torch.distributed.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/base.py]: clean the code

File: xtuner/v1/model/dense/dense.py [hunk 9]
```diff
@@ -144,7 +145,8 @@ def from_hf(self, hf_path: str | Path, strict: bool = True) -> tuple:
         loaded_keys, unloaded_keys, missing_keys = super().from_hf(hf_path, strict)
         # If model is built on meta device, we need to rebuild rotary embedding since from_hf will not
         # load the `inv_freq` of RotaryEmbedding which is a inpersisitent buffer.
-        self.rotary_emb = self.build_rotary_embedding(self.config)
+        # xTODO: remove this line below when with torch.device(DEVICE) in __init__()
+        # self.rotary_emb = self.build_rotary_embedding(self.config)
```
Comment (HAOCHENYE) [inline xtuner/v1/model/dense/dense.py:149]:
```suggestion
        # TODO: remove this line below when with torch.device(DEVICE) in __init__()
        # self.rotary_emb = self.build_rotary_embedding(self.config)
```

File: xtuner/v1/model/dense/dense.py [hunk 10]
```diff
@@ -182,7 +184,8 @@ def fully_shard(
             for param in self.parameters():
                 param.requires_grad = False
 
-        self.rotary_emb = self.build_rotary_embedding(self.config)
+        # xTODO: remove this line below when with torch.device(DEVICE) in __init__()
+        # self.rotary_emb = self.build_rotary_embedding(self.config)
```
Comment (HAOCHENYE) [inline xtuner/v1/model/dense/dense.py:188]:
```suggestion
        # TODO: remove this line below when with torch.device(DEVICE) in __init__()
        # self.rotary_emb = self.build_rotary_embedding(self.config)
```

File: xtuner/v1/model/moe/moe.py [hunk 11]
```diff
@@ -493,7 +494,9 @@ def _forward(
 
         # create position embeddings to be shared across the decoder layers
         assert position_ids is not None
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: clean the code

File: xtuner/v1/model/moe/moe.py [hunk 12]
```diff
@@ -493,7 +494,9 @@ def _forward(
...
         position_embeddings = self.rotary_emb(hidden_states, position_ids)
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: clean the code

File: xtuner/v1/model/moe/moe.py [hunk 13]
```diff
@@ -636,10 +639,17 @@ def build_rotary_embedding(self, config: MoEConfig) -> RotaryEmbeddingProtocol:
 
     @override
     def from_hf(self, hf_path: str | Path, strict: bool = True) -> tuple:
-        loaded_keys, unloaded_keys, missing_keys = super().from_hf(hf_path, strict)
         # If model is built on meta device, we need to rebuild rotary embedding since from_hf will not
         # load the `inv_freq` of RotaryEmbedding which is a inpersisitent buffer.
-        self.rotary_emb = self.build_rotary_embedding(self.config)
+        # This is used for training without FSDP.
+        # TODO: remove this because init with actual device already, Check it
+        # 对于FoPE, rotary_emb参数，如sin_coef，在之前full_shard已经被切分，但是这里重新构建，会让sin_coef又变回切分前。导致问题，如 load时尺寸不一致等。
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Keep english comments as above

File: xtuner/v1/model/moe/moe.py [hunk 14]
```diff
@@ -686,7 +696,10 @@ def fully_shard(
         if self.ep_mesh.size() > 1:
             self._replicate_other_params(self)
 
-        self.rotary_emb = self.build_rotary_embedding(self.config)
+        # 虽然 __init__ 中已经构建了 rotary_emb，但是是在 meta device 中构建的，这里需要在真实 device 上重新构建，
```
Comment (HAOCHENYE) [inline xtuner/v1/model/moe/moe.py]: Same

File: xtuner/v1/module/rope/rope.py [hunk 15]
```diff
@@ -115,6 +131,134 @@ def __call__(  # type: ignore
...
+        self.num_inv_freq = rope_scaling_cfg.num_inv_freq
+        self.fope_sep_head = rope_scaling_cfg.fope_sep_head
+        self.fope_init_factor = rope_scaling_cfg.fope_init_factor
+        if self.num_inv_freq is not None:
+            assert (self.inv_freq > (2.0 * torch.pi / config.max_position_embeddings)).all() or (
+                self.inv_freq.shape[-1] == self.num_inv_freq
+            ), "FoPE is wrongly initialized."
+
+        # zero out under-trained frequencies
+        # TODO: 单测
```
Comment (HAOCHENYE) [inline xtuner/v1/module/rope/rope.py]: Keep english comment as above

File: xtuner/v1/module/rope/rope.py [hunk 16]
```diff
@@ -40,6 +53,8 @@ def forward(self, x: torch.Tensor, position_ids: torch.LongTensor) -> tuple[torc
 
     def __call__(self, x: torch.Tensor, position_ids: torch.LongTensor) -> tuple[torch.Tensor, torch.Tensor]: ...
 
+    def to(self, device: torch.device) -> "RotaryEmbeddingProtocol": ...
```
Comment (HAOCHENYE) [inline xtuner/v1/module/rope/rope.py]: Should this return annotation be a forward reference?

File: xtuner/v1/ops/rotary_emb.py [hunk 17]
```diff
@@ -44,6 +49,43 @@ def apply_rotary_pos_emb_cuda(
     return q_embed, k_embed
 
 
+def apply_rotary_pos_emb_sep_cuda(q, k, cos, sin, position_ids=None, unsqueeze_dim=1):
```
Comment (HAOCHENYE) [inline xtuner/v1/ops/rotary_emb.py]: Miss typehint

File: xtuner/v1/ops/rotary_emb.py [hunk 18]
```diff
@@ -71,11 +113,16 @@ def __call__(
     ) -> Tuple[torch.Tensor, torch.Tensor]: ...
 
 
-def get_apply_rotary_emb() -> ApplyRotaryEmbProtocol:
+def get_apply_rotary_emb(fope_sep_head: bool | None = None) -> ApplyRotaryEmbProtocol:
```
Comment (HAOCHENYE) [inline xtuner/v1/ops/rotary_emb.py]: Since it's a bool type, why is its default value None?

File: xtuner/v1/train/trainer.py [hunk 19]
```diff
@@ -436,6 +436,7 @@ def __init__(
             strict=strict_load,
             intra_layer_micro_batch=intra_layer_micro_batch,
         )
+        # torch.distributed.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/train/trainer.py]: clean the code

File: xtuner/v1/train/trainer.py [hunk 20]
```diff
@@ -793,6 +794,7 @@ def build_engine(
 
         if model_path is not None:
             engine.model.set_hf(model_path)
+        # dist.breakpoint()
```
Comment (HAOCHENYE) [inline xtuner/v1/train/trainer.py]: clean the code

File: xtuner/v1/model/dense/dense.py (hunk 9 above)
Comment (jayhenry) [reply xtuner/v1/model/dense/dense.py:149]: xTODO表示把TODO已经完成了，跑过了单测

File: tests/model/test_qwen3_moe.py (hunk 1 above)
Comment (jayhenry) [reply tests/model/test_qwen3_moe.py:468]: 老版本的transformer存储的torch_dtype，但是新版本存的dtype，来兼容这个逻辑

File: tests/module/test_rope.py
Comment (jayhenry) [reply tests/module/test_rope.py:1]: 这个和rope.py文件对应，另外FoPE也是一种RoPE的变体，他的forward等接口和用法与RoPE相同

File: tests/model/test_qwen3_moe.py (hunk 2 above)
Comment (jayhenry) [reply tests/model/test_qwen3_moe.py:362]: OK

**Review by HAOCHENYE — APPROVED** [review]
